Implement paste

armenic opened the pull request:
Hi Doug,

I finally had a chance to implement paste (closes #13) 😄 I will appreciate you time and feedback 🙏  Please note the following:
- paste supports ✔️ :
  - `sep`
  - `collapse`
- paste does not support ⚠️:
  - `recycle0` since I could not understand its meaning, maybe after some time it will click 🤣 
  - lists and gracefully informs about that to the user. I think we can revisit this after we implement `as.character(list())`
- included several tests, mostly inspired by R's native examples
- I moved builtins into a separate directory since `paste` and its tests were already making builtins very busy.
- For R added
  -  `as_character` , similar to other casts. It helped me to easily cast from other types into character.
  - `get_vec_string` along with `Into<Vec<String>>` for Vector to help converting character vectors into Rust String vector. I know that normally if `From<Vec<String>>` is implemented then automatically `into` should work and I could not make it to work until I implemented into as well.
- Fixed the #17 
- Whenever we get 0 length vector I noticed that it is displayed as `numeric(0)`, you can reproduce it with `paste(NULL)`. It is caused by `src\r_vector\vectors.rs`, not by `paste()`:
```R
            if n == 0 {
                return write!(f, "numeric(0)");
            }
```

dgkf commented:
Thanks @armenic! Looks great after a quick look. This is quite a lot of code so it will take me a bit of time to review. There are a few other projects that are eating up my time right now, but I'll make my way to this and get a comprehensive review in before too long.

dgkf commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,360 @@
+                    }
+                }
+            }
+            _ => stack_vals.push((k, v)),
```
There are _a lot_ of ways we could implement this, and I think this already gets us pretty far. I'll leave it up to you how far you want to take it, but I'll just mention what I see as areas for improvement. 

1. **Minimizing allocations**
    This code currently allocates a lot of intermediate vectors. Are these really needed? 
    In fact, if we know the length of all the vectors to be pasted together, then we should be able to determine the output length upfront. Furthermore, we shouldn't need to reallocate anything to build the new strings - we can just refer back to the input values. 
2. **Preallocate string buffer for collapse**
    In the case when collapse is non-NULL, we _should_ be able to calculate the total output length! We would then _only_ require that we allocate a new string with the appropriate length and populate it with the relevant pieces.

armenic replied:
Fully agree Doug and I excluded all the unnecessary extra loops. In essence we now have one loop to deal with function arguments and we do a lot of things in that one loop. The other loop is the main logic of paste with recycling and formatting the string.

The collapse is done with rust's join method. Do you think is not appropriate to join strings?

armenic replied:
This already covered with Doug's nice refactor.

dgkf commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,360 @@
+            }
+            "collapse" => {
+                collapse = match v {
+                    R::Vector(Vector::Character(s_v)) => s_v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "collapse parameter must be a character string!".to_string(),
+                        )))
+                    }
```
Instead of operating on the raw internals, working with the higher-level R interface will help to perform coercion between data types for you so that you don't need to try to handle all sorts of edge cases. (untested code, but hopefully enough to set you off in the right direction)

```suggestion
                    R::Vector(v) => R::Vector(v.as_character()),
                    R::Null => R::Null,  // note this is a valid value for collapse
                    _ => {
                        return Err(RSignal::Error(RError::Other("invalid 'collapse' argument".to_string())))
                    }
```

armenic replied:
We need to check whether the supplied sep value is a character. R does not accept non-character values. If we use
`R::Vector(v) => R::Vector(v.as_character())` then this will coerce non-valid sep value to a character.

dgkf replied:
In the future I think it would be nice to allow anything that can be coerced to character, but let's start with matching R. This looks good.

dgkf commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,360 @@
+        // R c(1.1, 2, FALSE, "a", c(1, 2), sep = "+")
+        let mut keys = vec![None; 6];
+        keys.push(Some("sep".to_string()));
+        let args = ExprList {
+            keys,
+            values: vec![
+                Expr::Number(1.1),
+                Expr::Null,
+                Expr::Integer(2),
+                Expr::Bool(false),
+                Expr::String("a".to_string()),
+                Expr::Call(
+                    Box::new(Expr::Symbol("c".to_string())),
+                    ExprList {
+                        keys: vec![None; 2],
+                        values: vec![Expr::Number(1.0), Expr::Number(2.0)],
+                    },
+                ),
+                // sep parameter
+                Expr::String("+".to_string()),
+            ],
+        };
```
Allowing parsing will definitely make this a lot easier. If we get further into testing, I think we should add something like

```rust
let args: ExprList = parse("1.1, null, 2, false, \"a\", c(1.0, 2 = 2.0), sep = \"+\"");
```

dgkf replied:
Yeah, I didn't mean to say that I think it will work as-is, but I think we might want to add parsers for these sorts of partial expressions. I think we could have something like (similar to [this parse function](https://github.com/dgkf/R/blob/3b4ecad8d914cea55c0899abf408eabd9954aa6f/src/parser.rs#L39C18-L39C18))

```rust
pub fn parse_args(s: &str) -> Result<ExprList, RError> {
    match RParser::parse(Rule::repl, s) {
        Ok(pairs) => Ok(parse_pairlist(pairs)),
        Err(e) => Err(RError::ParseFailureVerbose(e)),
    }
}
```

Which will return an `ExprList` directly. One of the perks of having access to all the internal parsers at each stage of the syntax tree is that we can expose these helpers for testing if we need them.

armenic replied:
Thanks so much Doug 🙏 
This worked very nicely and thanks for enabling single quotes 💯

dgkf commented on src/r_vector/vectors.rs:
```diff
@@ -279,6 +279,21 @@ impl From<Vec<String>> for Vector {
 
+impl Into<Vec<String>> for Vector {
+    fn into(self) -> Vec<String> {
+        match self {
```
We can support more vector types by automatically doing the coercion before extracting string representations.

```suggestion
        match self.as_character() {
```

dgkf commented on src/r_vector/vectors.rs:
```diff
@@ -279,6 +279,21 @@ impl From<Vec<String>> for Vector {
+                    OptionNA::NA => "NA".to_string(),
+                })
+                .collect(),
+            _ => todo!(),
```
If we already tried to coerce to `character`, then we can replace this with 

```suggestion
            _ => unreachable!(),
```

dgkf reviewed (changes requested):
Really nice work! There are a couple test cases that we should aim to support before I'd feel comfortable pulling this in. At that point, there are a few performance or organizational changes that I suggested, but I'm fine with those being future work after we get an initial version in.

Specifically, these cases currently throw errors:

```r
x <- "<collapse>"
paste("a", "b", collapse = x)  # should allow variables as values
```

```r
paste("a", "b", collapse = NULL)  # should act as though empty
```

armenic commented:
Thanks so much Doug for your time reviewing and providing excellent feedback 🙏
I will do my best to address all your suggestions by the end of this week 🤞

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,360 @@
+    let collapse_is_specified = &args
+        .keys
+        .iter()
+        .position(|k| k == &Some("collapse".to_string()));
```
Great point Doug and we do not even need this. We now easily check whether the collapse has greater than zero length.

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,360 @@
+
+    let mut stack_vals: Vec<(_, _)> = vec![];
+    let mut sep = " ".to_string();
+    let mut collapse = String::new();
```
Excellent point Doug. I tried my best to make something useful and did not succeed. It turned out that have to do an additional loop to extract (or pop) the desired named arguments. Currently with one one loop we extract the desired named arguments, properly address them. In the same loop we also deal with the rest of arguments.

I will need your and other more experienced rustaceans' help to create `get_named()` 🙏

dgkf replied:
I think the way you have it now is probably the right path. There are a few learnings we can take away from the various thorny bits that you encountered here to hopefully push some of these things into helpers (like pop-ing out specific named args from the arg list)

armenic commented:
Doug, thanks for your patience and sorry for my beginner level mistakes :(

Your identified scenarios are now addressed and I included a test for the null value. I wanted to include a test with a vector and unfortunately because of #23 I could not create a vector and pass it to pated within the test. Otherwise it works nicely in repl.

I did a lot of refactor, simplified the code, factored out the `force_closures()` that you nicely used in `c` and reused it `paste()`. For the rest of the items and commented in the other threads.

Thanks for your time and help 🙏

dgkf commented:
> <sup>@armenic in [comment](https://github.com/dgkf/R/pull/19#issuecomment-1677466248)</sup>
> unfortunately because of #23 I could not create a vector

`;` should be parsed now in the REPL! I'm not sure about the new testing parser compatibility - it might depend on exactly how you plan to use it. Try it out and let me know if you're able to implement the tests that you want to use now.

armenic commented:
> `;` should be parsed now in the REPL! I'm not sure about the new testing parser compatibility - it might depend on exactly how you plan to use it. Try it out and let me know if you're able to implement the tests that you want to use now.

Thank you so much Doug for enabling the `;` and it works perfectly in REPL 💯 
I tried parsing it in the test, then I realized that the x variable needs to be available in the environment so that our primitives can evaluate their arguments. In `test_primitive_paste_09` I ended up inserting the x variable into env for testing purposes. Hopefully it is not too unorthodox way.

Hope you have a nice time off and certainly there is no rush reviewing this. 🙏

dgkf reviewed (commented):
Looking good! 

I suggested a minor refactor, but behaviorally it's identical. It's mostly just a matter of code style, so let me know what you think.

armenic commented on src/lang.rs:
```diff
@@ -138,6 +146,13 @@ impl R {
             _ => todo!(),
         }
     }
+
+    pub fn get_vec_string(&self) -> Vec<String> {
+        match self {
+            R::Vector(v) => v.clone().into(),
+            _ => todo!(),
+        }
+    }
```
Good catch Doug, thank you. Since now we have the into method we can use that in the test. Just destructure the `R::Vector` then use `.into` 💯. I will adjust the tests as well.

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,222 @@
+            .collect();
+    }
+
+    if collapse.len() > 0 {
+        output = vec![output.join(&collapse)];
+    }
```
Makes sense, thanks Doug.

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,222 @@
+
+    let mut output = vec!["".to_string(); max_len];
+
+    for i in 0..vec_s_vec.len() {
+        output = output
+            .iter()
+            // Any shorter vector will re-cycle its values to the length of
+            // longest one
+            .zip(vec_s_vec[i].iter().cycle())
+            .map(|(x, y)| {
+                // No need for a sep in the beginning
+                if i == 0 {
+                    return y.clone();
+                }
+                format!("{x}{sep}{y}")
+            })
+            .collect();
+    }
```
You are genius Doug, thanks so much. If I understand correctly creation of a new string from components is costlier than the pushing, right?

Very smart use of remainder operator 👍 The only thing we need to consider is the null. It is causing division by 0. I will fix this after accepting your suggestion, no worries. 

This implementation actually caught a bug where we were not using the sep when one of the items was null. I checked that R `paste(1, NULL, 2)` also injects the sep. I will adjust our tests that have the null case.

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,222 @@
+    let vals = force_closures(vals);
+
+    let mut vec_s_vec: Vec<Vec<String>> = vec![];
+    let mut sep = String::from(" ");
```
Indeed, thanks so much Doug. I will add this test.

armenic commented on src/r_builtins/paste.rs:
```diff
@@ -0,0 +1,222 @@
+                }
+            }
+
+            "collapse" => {
+                collapse = match v {
+                    R::Null => continue,
+                    R::Vector(Vector::Character(v)) => v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "collapse parameter must be NULL or a character string!".to_string(),
+                        )))
+                    }
+                }
+            }
+
+            _ => match v {
+                R::List(_) => {
+                    return Err(RSignal::Error(RError::Other(
+                        "list is not supported in paste() yet!".to_string(),
+                    )))
+                }
+
+                R::Null => continue,
+
+                _ => {
+                    // Leave the rest for processing. Coerce everything into
+                    // character.
+                    let s_vec = v.clone().as_character().unwrap().get_vec_string();
+                    vec_s_vec.push(s_vec.clone());
+
+                    // Need the length of longest vector to create an empty
+                    // vector that others will go through and re-cycle values as
+                    // needed
+                    if s_vec.len() > max_len {
+                        max_len = s_vec.len()
+                    }
+                }
+            },
+        }
+    }
```
Fully agreed, thanks so much. I was afraid of using the `remove` since its documentation says:
```
Note: Because this shifts over the remaining elements, it has a worst-case performance of O(n).
```

But I guess we don't need to worry too much about this since most of users just supply few inputs.

armenic commented:
Thanks so much Doug for fantastic refactor and and finding couple of bugs 🙏 
All your suggestions are implemented and I added the null ignoring part, adjusted couple of tests and added a new one.

dgkf commented:
Great! Looks good to me! Thanks for implementing this and iterating a bit.
